nixos/lib/test-driver/lib/machine.py

```diff
@@ -0,0 +1,868 @@
+        + "-sharpen 1 -posterize 3 -negate -gamma 100 "
+        + "-blur 1x65535"
+    )
+
+    tess_args = f"-c debug_file=/dev/null --psm 11"
+
+    cmd = f"convert {magick_args} {screenshot_path} tiff:{screenshot_path}.tiff"
+    ret = subprocess.run(cmd, shell=True, capture_output=True)
+    if ret.returncode != 0:
+        raise Exception(f"TIFF conversion failed with exit code {ret.returncode}")
+
+    model_results = []
+    for model_id in model_ids:
+        cmd = f"tesseract {screenshot_path}.tiff - {tess_args} --oem {model_id}"
+        ret = subprocess.run(cmd, shell=True, capture_output=True)
+        if ret.returncode != 0:
+            raise Exception(f"OCR failed with exit code {ret.returncode}")
+        model_results.append(ret.stdout.decode("utf-8"))
+
+    return model_results
+
+
+class BaseStartCommand:
+    def __init__(self):
+        pass
+
+    def cmd(
+        self,
+        monitor_socket_path: Path,
+        shell_socket_path: Path,
+        allow_reboot: bool = False,  # TODO: unused, legacy?
+        tty_path: Optional[Path] = None,  # TODO: currently unused
+    ):
+        tty_opts = ""
+        if tty_path is not None:
+            tty_opts += (
+                f" -chardev socket,server,nowait,id=console,path={tty_path}"
+                " -device virtconsole,chardev=console"
+            )
+        display_opts = ""
+        display_available = any(
+            x in os.environ for x in ["DISPLAY", "WAYLAND_DISPLAY"])
+        if display_available:
+            display_opts += " -nographic"
+
+        # qemu options
+        qemu_opts = ""
+        qemu_opts += (
+            "" if allow_reboot else " -no-reboot"
+            " -device virtio-serial"
+            " -device virtconsole,chardev=shell"
+            " -serial stdio"
+        )
+        # TODO: qemu script already catpures this env variable, legacy?
+        qemu_opts += " " + os.environ.get("QEMU_OPTS", "")
+
+        return (
+            f"{self._cmd}"
+            f" -monitor unix:{monitor_socket_path}"
+            f" -chardev socket,id=shell,path={shell_socket_path}"
+            f"{tty_opts}{display_opts}"
+        )
+
+    def build_environment(
+        state_dir: Path,
+        shared_dir: Path,
+    ):
+        return dict(os.environ).update(
+            {
+                "TMPDIR": state_dir,
+                "SHARED_DIR": shared_dir,
+                "USE_TMPDIR": "1",
+            }
+        )
+
+    def run(
+        self,
+        state_dir: Path,
+        shared_dir: Path,
+        monitor_socket_path: Path,
+        shell_socket_path: Path,
+    ):
+        return subprocess.Popen(
+            self.cmd(monitor_socket_path, shell_socket_path),
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            shell=True,
+            cwd=state_dir,
+            env=self.build_environment(state_dir, shared_dir),
+        )
+
+
+class NixStartScript(BaseStartCommand):
```

Comment:
unfortunately i don't see why this should not be a free function, when there is basically only one place where it could be called once and then the return value would be put into the name field of machine.

Reply:
It's a special property of the nix script that the name can be extracted the way we do here.

It would not necessarily apply to any other such start command.

Having this as property makes the calling code simpler and more parallel to the vlans calling code.

I highly value analogeous code, so I had no choice but to factore this into the class.

But since this particular extraction is a special (logical) property of the nix script only, it actually makes sense.

Reply:
The test driver class has all the domain specific oversight that we are starting nix-related scripts where you can parse the actual machine name out of the command name. therefore i think that the `machine_name` function could be a local function in the driver module so that it can instantiate `Machine` instances right from the start with a correct name, so we don't need any class hierarchies to model that. does this make sense?

Reply:
> does this make sense?

Not really, since the test driver is not the script domain (but the driver domain). And it _should_ not have that oversight. It's mixing domains. The driver domain should not care whether it's a nix script or anything else.

Does that make sense, in turn? :wink:

Reply:
We can continue on https://matrix.to/#/#nixpkgs-pr-125372:matrix.org if need be and unresolve in case it's need be. Marking as resolved for now since I don't plan to change anything, at the moment.